### PR TITLE
Add a CI and a basic integration test

### PR DESCRIPTION
**Pull request by Res260**
I had a burst of motivation and did this, I don't really care if you want it or not but offer it nonetheless, I mostly did this as a personnal challenge :)

**Basically, I added a Github Actions workflow that does two things on push and on Pull Request on the repo:**
**1. Try to build the Docker Image.**
**2. Try to install PyRDP and run a new integration test I made on ubuntu and windows.**
**3. Show a coverage report and fail it it falls under 50%**

This test is based on an RDP session I just recorded which try to use as most RDP features that PyRDP supports as possible. 

Right now, it does assertions on these:
- Asserts that PyRDP correctly build the requested virtual channels;
- Asserts that PyRDP sees a clipboard PDU with the pasted data;
- Asserts that the number of mouse clicks parsed matches the number of total clicks;
- Asserts that PyRDP clearly sees some text that was typed on the keyboard at the end of the session;

Beyond those assertions, it executes a large portion of the PyRDP codebase:
image
image

It does **NOT** test these:
- TLS, RC4 and anything else crypto
- Logging
- Replays
- Device Redirection Channel
- Executables in the "bin" folders
- The first two packets of the RDP connection

This test could be better (and can be improved in the future), but I believe it will catch most low-hanging fruits/bugs! 

Because of (what I believe is) a synchronization problem, I disabled testing for the device redirection channel.

It is heavily based on the pyrdp-replay branch (requires scapy as well).

### CODE REVIEW

File: test/test.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                test_mitm.recvFromClient(data)
+            elif source == server_ip and destination == mitm_ip:
+                test_mitm.recvFromServer(data)
+            elif source == mitm_ip and destination == client_ip and destination_port != 3389:
+                test_mitm.sendToClient(data)
+            elif source == mitm_ip and destination == server_ip:
+                test_mitm.sendToServer(data)
+            else:
+                assert False
+        except NotImplementedError as e:
```
Comment (alxbl) [inline test/test.py]: Isn't this entire try-block redundant? Just let the exception be thrown, unless I'm missing something?

File: test/test.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+            raise e
+
+    try:
+        test_mitm.tcp.recordConnectionClose()
+    except struct.error as e:
```
Comment (alxbl) [inline test/test.py]: Should this cause the test to fail?
Should you catch a generic exception? What if the code changes and throws a different exception? The test will no longer report failure/unexpected behavior

File: test/test.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+        print("Couldn't close the connection cleanly. "
+              "Are you sure you got source and destination correct?")
+
+    assert test_mitm.builtIOChannel, "PyRDP did not build IO Channel."
+    assert test_mitm.builtCliprdrChannel, "PyRDP did not build the Clipboard Channel."
+    assert test_mitm.builtRDPDRChannel, "PyRDP did not build the RDPDR Channel."
+
+    logging.critical("Channel building assertions PASSED")
```
Comment (alxbl) [inline test/test.py]: Don't need to log success in integration tests.
Also don't increase severity to ensure something gets printed. This should be INFO, if logged at all. In tests, no output == no problem

File: test/test.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+    for key, value in test_mitm.clipboardObserver.expectedClipboardData.items():
+        assert value, f"Expected to receive {key} in a clipboardPDU but the clipboard observer did not receive it."
+
+    logging.critical("Clipboard content assertions PASSED")
```
Comment (alxbl) [inline test/test.py]: See above

File: test/test.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    actualLeftMouseClicks = test_mitm.inputObserver.leftMouseClicks
+    expectedRightMouseClicks = 3
+    actualRightMouseClicks = test_mitm.inputObserver.rightMouseClicks
+    expectedMouseWheelClicks = 0
+    actualMouseWheelClicks = test_mitm.inputObserver.mouseWheelClicks
+    assert actualLeftMouseClicks == expectedLeftMouseClicks, f"Wrong number of left mouse clicks registered. Expected {expectedLeftMouseClicks}, got {actualLeftMouseClicks}"
+    assert actualRightMouseClicks == expectedRightMouseClicks, f"Wrong number of right mouse clicks registered. Expected {expectedRightMouseClicks}, got {actualRightMouseClicks}"
+    assert actualMouseWheelClicks == expectedMouseWheelClicks, f"Wrong number of mouse wheel clicks registered. Expected {expectedMouseWheelClicks}, got {actualMouseWheelClicks}"
+
+    logging.critical("Mouse clicks assertions PASSED")
```
Comment (alxbl) [inline test/test.py]: See above

File: test/test.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+    assert "arrrray" in test_mitm.state.inputBuffer, f"'arrrray' not found in the MITM state input buffer, but was typed during the session. Input buffer: '{test_mitm.state.inputBuffer}'"
+
+    logging.critical("Keyboard typing assertion PASSED")
```
Comment (alxbl) [inline test/test.py]: See above

**Comment by alxbl** [comment]
This is very cool! I think it's a great first step towards having proper unit testing and integration testing in PyRDP. I have a few comments but the general approach looks good to me.

Good use of dependency injection for the configs, too.

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    - name: Pip version
+      run: pip --version
+
+    - name: Install setuptools
+      run: sudo apt install python3-setuptools
+    - name: Install PyRDP dependencies
+      run: sudo apt install libdbus-1-dev libdbus-glib-1-dev libgl1-mesa-glx git python3-dev
+    - name: Install PyRDP
+      working-directory: .
+      run: pip install -U -e .
```
[the remark this answers is not in the document]
Comment (Res260) [reply .github/workflows/ci.yml]: It should be fine!

File: test/test.py (hunk 1 above)
Comment (Res260) [reply test/test.py]: Yeah, probably, it's from the pyrdp-replay script and I didnt remove it but probably should

File: test/test.py (hunk 3 above)
Comment (Res260) [reply test/test.py]: Makes sense, will change

File: test/test.py (hunk 2 above)
Comment (Res260) [reply test/test.py]: removed it

File: test/test.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    def __init__(self):
+        super().__init__()
+        self.leftMouseClicks = 0
+        self.rightMouseClicks = 0
+        self.mouseWheelClicks = 0
+
+    def onPDUReceived(self, pdu: FastPathPDU):
+        for event in pdu.events:
+            if isinstance(event, FastPathMouseEvent):
+                if event.pointerFlags & PointerFlag.PTRFLAGS_DOWN and event.pointerFlags & PointerFlag.PTRFLAGS_BUTTON1:
```
[the remark this answers is not in the document]
Comment (Res260) [reply test/test.py]: done

**Comment by alxbl** [comment]
If you can rebase this and `--force-with-lease` to clean up the commit history, I'd be willing to merge this now.

**Comment by Res260** [comment]
@alxbl squash and merge would do the job?

**Comment by obilodeau** [comment]
@alxbl I would like to review before you merge, please. I'm very interested in GitHub Actions beyond just this repo.

**Comment by alxbl** [comment]
> @alxbl wouldn't squash and merge do the job?
> 
> Edit: wording as the first message seemed passive-agressive

It will, but your commits will disappear. I was offering you to keep your commit history in master.

**Comment by Res260** [comment]
Ahhh I don't really care about having multiple commits for this PR, and the branch is still kept in GitHub if we really need the information

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+jobs:
+  docker-build:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: Build Docker image
+      run: docker build .
+
+  Install-and-test:
```
Comment (obilodeau) [inline .github/workflows/ci.yml]:
```suggestion
  install-and-test:
```
for consistency with other names

File: pyrdp/mitm/BasePathMITM.py [hunk 10]
```diff
@@ -60,4 +60,4 @@ def onScanCode(self, scanCode: int, isReleased: bool, isExtended: bool):
         # Normal input
         elif len(keyName) == 1:
             if not isReleased:
-                self.state.inputBuffer += keyName
\ No newline at end of file
+                self.state.inputBuffer += keyName
```
Comment (obilodeau) [inline pyrdp/mitm/BasePathMITM.py]: Please refrain from unnecessary line-noise. Learn to use `git add -p` or your IDE's equivalent.

File: pyrdp/mitm/RDPMITM.py [hunk 11]
```diff
@@ -37,15 +37,16 @@
...
+    Recorder
 
 
 class RDPMITM:
     """
     Main MITM class. The job of this class is to orchestrate the components for all the protocols.
     """
 
-    def __init__(self, mainLogger: SessionLogger, crawlerLogger: SessionLogger, config: MITMConfig):
+    def __init__(self, mainLogger: SessionLogger, crawlerLogger: SessionLogger, config: MITMConfig, state: RDPMITMState=None, recorder: Recorder=None):
```
Comment (obilodeau) [inline pyrdp/mitm/RDPMITM.py]: Note to self: this is duplicated from another active PR

File: test/test.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+from pyrdp.mitm.MITMRecorder import MITMRecorder
+from pyrdp.mitm.state import RDPMITMState
+from pyrdp.pdu import ClipboardPDU, FastPathMouseEvent, FastPathPDU, FormatDataResponsePDU
+
+
+def bytesToIP(data: bytes):
+    return ".".join(str(b) for b in data)
+
+
+def parseExportedPdu(packet: packet.Raw):
```
Comment (obilodeau) [inline test/test.py]: Code duplicated from an active PR. One of us will rebase and refactor to avoid this useless duplication.

**Comment by obilodeau** [comment]
A simple zip of the pcap gave a 5.9M binary (from 16M). It's not as good as I have expected. Since git stores it's pack metadata already compressed (so also on the network), I'm going to let go on this.

I might revisit if we add more pcap-based tests to reduce checkout size.

File: .github/workflows/ci.yml [hunk 13]
```diff
@@ -36,4 +36,4 @@ jobs:
 
     - name: Integration Test with a prerecorded PCAP.
       working-directory: ./
-      run: python test/test.py
\ No newline at end of file
+      run: python test/test.py
```
Comment (obilodeau) [inline .github/workflows/ci.yml]: again! ;)

**Comment by Res260** [comment]
> A simple zip of the pcap gave a 5.9M binary (from 16M). It's not as good as I have expected. Since git stores it's pack metadata already compressed (so also on the network), I'm going to let go on this.
> 
> I might revisit if we add more pcap-based tests to reduce checkout size.

Should I revert the commit then?

File: pyrdp/mitm/BasePathMITM.py (hunk 10 above)
Comment (Res260) [reply pyrdp/mitm/BasePathMITM.py]: but pep8 :(

**Comment by obilodeau** [comment]
> Should I revert the commit then?

Since it's already done I'm tempted to say no.

That said, are you sure this directory will always be writable to the user running the tests? It looks like the GitHub action worked so we can keep it.

**Comment by Res260** [comment]
> That said, are you sure this directory will always be writable to the user running the tests? It looks like the GitHub action worked so we can keep it.

There is no garantee indeed, i'm not sure what would be the best approach. It will always work for the CI though

**Comment by Res260** [comment]
For the code coverage, I could add it to the CI if you want using https://www.codecov.io/, thoughts?

File: .github/workflows/ci.yml (hunk 13 above)
Comment (Res260) [reply .github/workflows/ci.yml]: Not sure I understand 😂

**Comment by Res260** [comment]
Actually I can do it inside the CI without a third party

File: pyrdp/mitm/BasePathMITM.py (hunk 10 above)
Comment (obilodeau) [reply pyrdp/mitm/BasePathMITM.py]: Pep8 says files should end with a `\n` (like vim does), I'm all for it.

Reducing line noise is about making it easier to review for the reviewer by removing all the noise that has nothing to do with the PR. If you touch a function or create a new file, please follow PEP8 but if you do PEP8 enforcement changes in the same commit / PR as another feature then it makes review harder to do.

When an existing code-base already has lacking PEP8 enforcement, using a tool to convert it all to PEP8 compliant is another case of noise that is added for little benefit. Most of the time when trying to find commit introducing problems with `git log -p` reverse search, you will always stumble on these big "re-indent" commits and it slows down your quest. For such codebases the re-indent / pep8 enforcement should be done gradually as you are modifying the code for another purpose.

Of course, this is a mix of experience and personal preference, your mileage may vary.

**Comment by obilodeau** [comment]
> For the code coverage, I could add it to the CI if you want using https://www.codecov.io/, thoughts?

then

> Actually I can do it inside the CI without a third party

Do what you prefer.

**Comment by Res260** [comment]
Added windows installation and test, as a treat :)

It does the same thing as the ubuntu one but on windows. My reasonning behind this is that even though installation does not fail per se, execution can fail because dependencies sometime have a hard time testing for both windows and linux, so I'd rather run the test twice.

**Comment by Res260** [comment]
I added the coverage report to the CI, and it will fail if it falls under 50% (currently 53% coverage).

@obilodeau Should I still add documentation for the coverage in docs or it's not needed?

**Comment by obilodeau** [comment]
> @obilodeau Should I still add documentation for the coverage in docs or it's not needed?

Not needed. This is self-explanatory now. Thanks!

**Pull request merged**